src/js/components/Button/Button.js

```diff
@@ -92,12 +96,80 @@ const getIconColor = (paths = [], theme, colorProp, kind) => {
   return result[1] || undefined;
 };
 
+const getBadgeDimension = (dimension, badgeProp, badgeRef, theme) => {
+  if (
+    typeof badgeProp === 'number' ||
+    (typeof badgeProp === 'object' && badgeProp.value)
+  ) {
+    const borderWidth = badgeProp.border
+      ? parseInt(
+          theme.global.borderSize[badgeProp.border.size].replace('px', ''),
+          10,
+        ) * 2
+      : 0;
+    // leave a small amount of horizontal space to pad content
+    const horizontalPad =
+      dimension === 'width'
+        ? parseInt(theme.global.edgeSize.xsmall.replace('px', ''), 10)
+        : 0;
+    // if content is tall/wide, let badge grow to fit. otherwise,
+    // make sure it's at least badge.size.medium dimensions
+    return `${Math.max(
+      Math.ceil(badgeRef.current.getBoundingClientRect()[dimension]) +
+        horizontalPad +
+        borderWidth,
+      parseInt(theme.button.badge.size.medium.replace('px', ''), 10) +
+        borderWidth,
+    )}px`;
+  }
+  return `${badgeRef.current.getBoundingClientRect()[dimension]}px`;
+};
+
+const getBadge = (badgeProp, badgeRef, badgeHeight, badgeWidth, theme) => {
+  const max = badgeProp.max || theme.button.badge.max;
```

Comment:
I'm not sure why `max` would be in the theme. Any reason to not just have a default value for the property?

Reply:
We could set a default value for the property. I guess my mindset was by default there would be no max, but a caller could specify in their theme that across their app they wanted a max. In Design System case, we would want max to be 9, so that anything greater would be 9+. If you're comfortable with grommet having a default max value, we could set max to 9 by default here.

Reply:
this feels like the right 80% default case